CNV Live migration of VMIs

aburdenthehand opened the pull request:
This PR covers the following stories:
CNV-757 (Basic live migration)
CNV-759 (Progress/monitoring)
CNV-761(Cancel migration)
CNV-1495 (UI for above) 

No changes to the topic map as yet as migrating 1.4 content and assembling the CNV User Guide will begin next week. 

Preview content hosted here:
http://file.bne.redhat.com/aburden/CNV-docs-preview/migration/cnv_users_guide/cnv-live-migration.html
http://file.bne.redhat.com/aburden/CNV-docs-preview/migration/cnv_users_guide/cnv-live-migration-limits.html
http://file.bne.redhat.com/aburden/CNV-docs-preview/migration/cnv_users_guide/cnv-migrate-vmi.html
http://file.bne.redhat.com/aburden/CNV-docs-preview/migration/cnv_users_guide/cnv-monitor-vmi-migration.html
http://file.bne.redhat.com/aburden/CNV-docs-preview/migration/cnv_users_guide/cnv-cancel-vmi-migration.html

suomiy commented on modules/cnv-cancelling-vm-migration-web.adoc:
```diff
@@ -0,0 +1,21 @@
+= Cancelling live migration of a virtual machine instance in the web console
+
+A live migration of the virtual machine instance can be cancelled using the 
+&#8942; button found on each virtual machine in the *Workloads* -> *Virtual Machines* 
```
kebap button (&#8942;) 

might be more descriptive

aburdenthehand replied:
The 'kebab/kebap' menu conversation came up in OCP docs and I'm waiting to hear what the decision was. 
I'd be more inclined to call it an 'Actions' menu, since it mostly mirrors the options available for that button on the <Object> Description page

suomiy replied:
+1 Action menu sounds good

suomiy commented on modules/cnv-configuring-vmi-live-migration-cli.adoc:
```diff
@@ -0,0 +1,42 @@
+Only custom virtual machines need to have this feature manually configured. This
+ eviction strategy is configured by default in virtual machine created from 
+common templates, or if a guest operating system is selected in either the 
+*Create Virtual Machine* or *Create Virtual Machine Template* wizards in the 
```
This is thanks to the templates - we do not configure eviction strategy explicitly

aburdenthehand replied:
Sorry @suomiy, I'm not sure I understand. Can you please elaborate?

suomiy replied:
The VM is always created from a template - so the eviction strategy is not because of the guest operating system.

aburdenthehand replied:
@suomiy I removed the part about guest operating systems. Does that make more sense for you?

suomiy replied:
yes, looks ok

dshchedr commented on modules/cnv-initiating-vm-migration-cli.adoc:
```diff
@@ -0,0 +1,34 @@
+----
+
+The *VirtualMachineInstanceMigration* object triggers a live migration of the 
+virtual machine instance and exists until the migration is complete. 
```
As I know, *VirtualMachineInstanceMigration* object will exist all time while VMI exists (it will be removed only after stopping/deleting VM)

vladikr replied:
Yes. Live migration object will remain as long as the VMI object exist or if manually deleted. In fact, there can be multiple migration object for each VMI.

aburdenthehand replied:
Updated to read: "The *VirtualMachineInstanceMigration* object triggers a live migration of the
virtual machine instance. This object exists in the cluster for as long as the
virtual machine instance is running, unless manually deleted."

vladikr commented on modules/cnv-configuring-live-migration-limits.adoc:
```diff
@@ -0,0 +1,40 @@
+// cnv_users_guide/cnv-live-migration-limits.adoc
+
+[id="cnv-configuring-live-migration-limits_{context}"]
+= Configuring live migration limits
```
I'm not sure about the word `limits`. There can be different options, not just limiting factors. That's minor though...

aburdenthehand replied:
I've added "and timeouts" where appropriate.

vladikr commented on modules/cnv-live-migration-limits-reftable.adoc:
```diff
@@ -0,0 +1,36 @@
+|`completionTimeoutPerGiB`
+|The migration will be aborted if it has not completed in this time, in seconds 
+per GiB of memory. For example, a VMI with 6GiB memory will timeout if it has 
+not completed migration in 4800 seconds.
```
In case of a block migration, the size of the disks that are being migrated will be included in the calculation.

aburdenthehand replied:
Now reads:
"The migration will be aborted if it has not completed in this time, in seconds
per GiB of memory. For example, a VMI with 6GiB memory will timeout if it has
not completed migration in 4800 seconds. If the `Migration Method` is
`BlockMigration`, the size of the migrating disks is included in the calculation."

dshchedr commented:
lgfm, thanks

vikram-redhat commented:
@openshift/team-documentation PTAL.

kalexand-rh commented on modules/cnv-initiating-vm-migration-web.adoc:
```diff
@@ -0,0 +1,27 @@
+
+Migrate a running virtual machine instance to a different node in the cluster. 
+
+Migrate a virtual machine using the &#8942; button of the template in the 
```
I don't think you need this paragraph.

kalexand-rh commented on modules/cnv-initiating-vm-migration-web.adoc:
```diff
@@ -0,0 +1,27 @@
+
+[NOTE]
+====
+Only admin users can initiate a virtual machine migration, although the 
```
I'd swap the clause order and consider making this a prereq instead of a note.

kalexand-rh commented on modules/cnv-initiating-vm-migration-web.adoc:
```diff
@@ -0,0 +1,27 @@
+.Procedure
+
+. In the {ProductName} console, click *Workloads* -> *Virtual Machines*.
+. Click the &#8942; button of the virtual machine to migrate and select 
```
My suggestions for these alternate paths are the same as in the other module: I'd pull one, but if you keep them both, please restructure the step.

kalexand-rh commented on modules/cnv-live-migration-limits-reftable.adoc:
```diff
@@ -0,0 +1,37 @@
+
+.Migration parameters
+|===
+|Parameter |Description |Default
```
Default value

If you have information about acceptable ranges, consider adding another column to describe the values.

kalexand-rh commented on modules/cnv-live-migration-limits-reftable.adoc:
```diff
@@ -0,0 +1,37 @@
+|64Mi
+
+|`completionTimeoutPerGiB`
+|The migration will be aborted if it has not completed in this time, in seconds 
```
not "abort"

> :no_entry_sign: abort v
Do not use. Use “cancel,” “stop,” or “end."

kalexand-rh commented on modules/cnv-live-migration-limits-reftable.adoc:
```diff
@@ -0,0 +1,37 @@
+
+|`completionTimeoutPerGiB`
+|The migration will be aborted if it has not completed in this time, in seconds 
+per GiB of memory. For example, a VMI with 6GiB memory will timeout if it has 
```
I'd mention the value of completionTimeoutPerGiB here. It's implied that it's the default value, but I've seen people file bugs about similar issues.

I'd also consider removing the future tense. I think it's clear enough in present tense.

kalexand-rh commented on modules/cnv-live-migration-limits-reftable.adoc:
```diff
@@ -0,0 +1,37 @@
+|800
+
+|`progressTimeout`
+|The migration will be aborted if memory copy fails to make progress in this 
```
not "abort"

kalexand-rh commented on modules/cnv-monitoring-vm-migration-cli.adoc:
```diff
@@ -0,0 +1,39 @@
+* Use the `oc describe` command on the migrating virtual machine instance:
+
+----
+oc describe vmi vmi-fedora
```
$

kalexand-rh commented on modules/cnv-monitoring-vm-migration-web.adoc:
```diff
@@ -0,0 +1,17 @@
+[id="cnv-monitoring-vm-migration-web_{context}"]
+= Monitoring live migration of a virtual machine instance in the web console
+
+For the duration of the migration, the virtual machine will have a status of 
```
s/will have/has

kalexand-rh commented on modules/cnv-monitoring-vm-migration-web.adoc:
```diff
@@ -0,0 +1,17 @@
+= Monitoring live migration of a virtual machine instance in the web console
+
+For the duration of the migration, the virtual machine will have a status of 
+*Migrating*. This is displayed in the *Virtual Machines* list, or in the 
```
s/This/This status

no comma

kalexand-rh commented on modules/cnv-monitoring-vm-migration-web.adoc:
```diff
@@ -0,0 +1,17 @@
+
+.Procedure
+
+. In the {ProductName} console, click *Workloads* -> *Virtual Machines*.
```
Same as before, I'd pick one, but if you keep them both, please reformat the steps.

Since there's one step, lead with * instead of .

aburdenthehand commented on modules/cnv-cancelling-vm-migration-web.adoc:
```diff
@@ -0,0 +1,21 @@
+.Procedure
+
+. In the {ProductName} console, click *Workloads* -> *Virtual Machines*.
+. Click the Actions &#8942; button of the virtual machine and select 
```
I think there's benefit to having both methods. How about:
. You can cancel the migration from this screen, which makes it easier to perform actions on multiple virtual machines in the one screen, or from the *Virtual Machine Details* screen:
** Click :kebab: at the end of  virtual machine and select           
*Cancel Virtual Machine Migration*.
** Click the virtual machine name to open the *Virtual Machine Details*               
screen and click *Actions* -> *Cancel Virtual Machine Migration*.
?
It's a bit wordy, but it offers a 'why?' for the first method - the second method should be self-explanatory.

kalexand-rh replied:
That works! It's good to have the reason why you'd pick the other one. I'd apply the same style to the other similar instances.

aburdenthehand commented on modules/cnv-configuring-live-migration-limits.adoc:
```diff
@@ -0,0 +1,40 @@
+oc edit configmap kubevirt-config -n kubevirt
+----
++
+[source,yaml]
```
The reference table is the next module in the assembly. I'd prefer a link as well but ...

aburdenthehand commented on modules/cnv-configuring-vmi-live-migration-cli.adoc:
```diff
@@ -0,0 +1,40 @@
+
+.Prerequisites
+
+* The live migration feature must be enabled in the cluster.
```
Right you are. This is for #14923 (node maintenance). I mistakenly added this instead of the live migration concept that is shared across both that PR and this one. I've fixed that now, and will add the feedback you've noted for this topic in the 14923.

aburdenthehand commented on modules/cnv-configuring-vmi-live-migration-cli.adoc:
```diff
@@ -0,0 +1,40 @@
+
+* Add the `evictionStrategy: LiveMigrate` to the `spec` in the virtual machine instance 
+configuration file. The following example snippet shows the eviction strategy 
+configured in the *_custom-vmi_* configuration file:
```
It's just an example VMI config so I've changed the procedure to use `$ oc edit vmi ...` for the example

aburdenthehand commented:
@kalexand-rh - Thanks for the review Kathryn. I have updated the PR as necessary.
I added a new file cnv-understanding-live-migration.adoc that is shared with another PR (as discussed). If it passes muster then I believe this is ready to merge.

kalexand-rh commented:
I scanned the final PR, and it looks fine. Merging.

kalexand-rh commented:
/cherrypick enterprise-4.1

kalexand-rh commented:
/cherrypick enterprise-4.2
